src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -1946,7 +1946,28 @@ class ArnoldLight : public ArnoldObject
 
 			// Update light shader.
 
+			// Delete current light shader, destroying all AtNodes it owns. It
+			// is crucial that we do this _before_ constructing a new
+			// ArnoldShader (and therefore AtNodes) below, because we are
+			// relying on a specific behaviour of the Arnold node allocator.
+			// When we destroy the light node, Arnold does not remove it from
+			// any of the `light_group` arrays we have assigned to geometry,
+			// meaning they will contain a dangling pointer. If we destroy the
+			// old AtNode first, we get lucky, and Arnold will allocate the new
+			// one at the _exact same address_ as the old one, keeping our
+			// arrays valid. We have been accidentally relying on this behaviour
+			// for some time, and for now continue to rely on it in lieu of a
+			// more complex fix. One subtlety is that when the light shader
```

Comment:
> One subtlety is that when the light shader contains more than one node, we must destroy them in reverse order - see ~ArnoldShader.

This isn't true for this branch is it?

Reply:
It isn't, sorry. I blindly cherry-picked the commit over. Will fix as soon as I get into work. Thanks for pointing that out.